File: next-js-boilerplate/src/lib/tier.ts
// Tier ordering mirrors the backend at
// nest-js-boilerplate/src/authorization/tier.enum.ts — keep in sync.
//
// The backend is the source of truth; this module is a frontend convenience
// for render-only gating (TierGate, useMinTier). Every gated request still
// hits @MinTier on the backend.

export const TIERS = ["FREE", "BASIC", "MEDIUM", "PREMIUM"] as const;
export type Tier = (typeof TIERS)[number];
export const TIER_ORDER: Record<Tier, number> = {
  FREE: 0,
  BASIC: 1,
  MEDIUM: 2,
  PREMIUM: 3,
};

export function tierAtLeast(
  userTier: string | null | undefined,
  min: Tier,
): boolean {
  if (!userTier) return false;
  const userRank = TIER_ORDER[userTier as Tier];
  if (userRank === undefined) return false;
  return userRank >= TIER_ORDER[min];
}

export function tierLabel(tier: string): string {
  return tier.charAt(0) + tier.slice(1).toLowerCase();
}

export const TIER_PRICES_CENTS: Record<Tier, number> = {
  FREE: 0,
  BASIC: 999,
  MEDIUM: 1999,
  PREMIUM: 4999,
};

File: flutter-boilerplate/lib/lib/tier.dart
class Tier {
  Tier._();

  static const String free = 'free';
  static const String basic = 'basic';
  static const String medium = 'medium';
  static const String premium = 'premium';

  static const List<String> all = [free, basic, medium, premium];

  static const tierOrder = {
    free: 0,
    basic: 1,
    medium: 2,
    premium: 3,
  };

  static bool hasAccess(String userTier, String requiredTier) {
    final userLevel = tierOrder[userTier] ?? 0;
    final requiredLevel = tierOrder[requiredTier] ?? 0;
    return userLevel >= requiredLevel;
  }

  static String displayName(String tier) {
    switch (tier) {
      case free:
        return 'Free';
      case basic:
        return 'Basic';
      case medium:
        return 'Medium';
      case premium:
        return 'Premium';
      default:
        return 'Free';
    }
  }
}
